### ISSUE TITLE
Filesize splitting questions

### ISSUE THREAD

**Issue by wmbutler** [qpdf/qpdf#203]
When splitting some native PDF's into smaller groups, I find that each group has a filesize close to the size of the original full document.

Is there a way to mitigate this? Seems that there must be a way to discard this information or compress it in some meaningful way.

**Comment by jberkenbilt** [comment]
Looks very similar to #174. The only way to resolve this would be to process the content stream. For the specific case of images, I think it would be possible to write something in qpdf that would do it with pretty good reliability, especially now with the new content filtering stuff introduced in version 8. Basically all the primitives exist in qpdf to make it possible to solve this. That was actually not true when I commented on the other issue.

**Comment by wmbutler** [comment]
@jberkenbilt I'm wondering if there is a generic way to solve this problem. Essentially, a way to recognize when images are linked improperly and correct them during the split process. This is something we would pay a bounty for because we see it in many cases for some large documents.

**Comment by jberkenbilt** [comment]
What I think is happening is this. The PDF pages dictionary allows certain resources to be inherited, so a common thing lazy PDF writers do is to put all the images at the top of the pages tree so they don't have to bother listing specific images with specific pages. This is completely valid and works fine in all readers, but it fakes out any software, including qpdf, that uses the metadata to map out which objects belong to which pages. What I would do would be to first push all the metadata down to the page level (a qpdf library routine to do this already exists), and then iterate through each page, scanning through the content stream for any name token that matches any XObject referenced in the metadata. For all XObjects that are not mentioned anywhere in the contents, I would simply filter them out of the metadata for that page. I suspect that this approach would be very reliable, and I could probably implement it in a couple of hours given the new APIs added in qpdf 8.

**Comment by wmbutler** [comment]
@jberkenbilt I would definitely pay a bounty for this.

**Comment by jberkenbilt** [comment]
This fix will be in 8.1, which I hope to get out in the next few days.

**Comment by jberkenbilt** [comment]
By the way, doing this fix really validated my recent changes. It only took me about 3 hours to implement this, though it would have been very hard to do this without the lexical changes I added for a completely different purpose, and it would also have been more cumbersome had it not been for the page helper APIs I added on Tuesday.

In any case, I spent 3 hours dedicated to fixing this specific issue.

**Comment by jberkenbilt** [comment]
By the way, if you'd like to test this, you can try the appimage: https://transfer.sh/jizsj/qpdf-git.6c89d4b-x86_64.AppImage

That should be there for something like two weeks, though I will have released 8.1 long before it disappears.

**Comment by jberkenbilt** [comment]
@mccool @wmbutler I'd estimate that I spent about 3 hours on this task.

The changes were straightforward. Most of the time was spent analyzing the code and exercising the code with different types of files (all permutations of use of object streams and linearization) to ensure that qpdf's estimate of progress would be reasonably accurate in all those cases and would advance in a reasonably uniform way as progress went on. I also spent time designing what the API should look like to make the feature useful to library users, not just users of the CLI.
